File: sdks/csharp/sdk/src/PogodocApi/Templates/Types/SaveCreatedTemplateRequestTemplateInfo.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using PogodocApi.Core;

namespace PogodocApi;

[Serializable]
public record SaveCreatedTemplateRequestTemplateInfo : IJsonOnDeserialized
{
    [JsonExtensionData]
    private readonly IDictionary<string, JsonElement> _extensionData =
        new Dictionary<string, JsonElement>();

    /// <summary>
    /// Title of the template
    /// </summary>
    [JsonPropertyName("title")]
    public required string Title { get; set; }

    /// <summary>
    /// Description of the template
    /// </summary>
    [JsonPropertyName("description")]
    public required string Description { get; set; }

    /// <summary>
    /// Type of template to be rendered
    /// </summary>
    [JsonPropertyName("type")]
    public required SaveCreatedTemplateRequestTemplateInfoType Type { get; set; }

    /// <summary>
    /// Sample data for the template
    /// </summary>
    [JsonPropertyName("sampleData")]
    public Dictionary<string, object?> SampleData { get; set; } = new Dictionary<string, object?>();

    [JsonPropertyName("sourceCode")]
    public string? SourceCode { get; set; }

    /// <summary>
    /// Categories of the template
    /// </summary>
    [JsonPropertyName("categories")]
    public IEnumerable<SaveCreatedTemplateRequestTemplateInfoCategoriesItem> Categories { get; set; } =
        new List<SaveCreatedTemplateRequestTemplateInfoCategoriesItem>();

    [JsonIgnore]
    public ReadOnlyAdditionalProperties AdditionalProperties { get; private set; } = new();

    void IJsonOnDeserialized.OnDeserialized() =>
        AdditionalProperties.CopyFromExtensionData(_extensionData);

    /// <inheritdoc />
    public override string ToString()
    {
        return JsonUtils.Serialize(this);
    }
}

File: sdks/php/src/client/src/Templates/Types/SaveCreatedTemplateRequestTemplateInfo.php
<?php

namespace Pogodoc\Templates\Types;

use Pogodoc\Core\Json\JsonSerializableType;
use Pogodoc\Core\Json\JsonProperty;
use Pogodoc\Core\Types\ArrayType;

class SaveCreatedTemplateRequestTemplateInfo extends JsonSerializableType
{
    /**
     * @var string $title Title of the template
     */
    #[JsonProperty('title')]
    public string $title;

    /**
     * @var string $description Description of the template
     */
    #[JsonProperty('description')]
    public string $description;

    /**
     * @var value-of<SaveCreatedTemplateRequestTemplateInfoType> $type Type of template to be rendered
     */
    #[JsonProperty('type')]
    public string $type;

    /**
     * @var array<string, mixed> $sampleData Sample data for the template
     */
    #[JsonProperty('sampleData'), ArrayType(['string' => 'mixed'])]
    public array $sampleData;

    /**
     * @var ?string $sourceCode
     */
    #[JsonProperty('sourceCode')]
    public ?string $sourceCode;

    /**
     * @var array<value-of<SaveCreatedTemplateRequestTemplateInfoCategoriesItem>> $categories Categories of the template
     */
    #[JsonProperty('categories'), ArrayType(['string'])]
    public array $categories;

    /**
     * @param array{
     *   title: string,
     *   description: string,
     *   type: value-of<SaveCreatedTemplateRequestTemplateInfoType>,
     *   sampleData: array<string, mixed>,
     *   categories: array<value-of<SaveCreatedTemplateRequestTemplateInfoCategoriesItem>>,
     *   sourceCode?: ?string,
     * } $values
     */
    public function __construct(
        array $values,
    ) {
        $this->title = $values['title'];
        $this->description = $values['description'];
        $this->type = $values['type'];
        $this->sampleData = $values['sampleData'];
        $this->sourceCode = $values['sourceCode'] ?? null;
        $this->categories = $values['categories'];
    }

    /**
     * @return string
     */
    public function __toString(): string
    {
        return $this->toJson();
    }
}
